import os
from PySide6.QtWidgets import QWidget, QMessageBox, QHeaderView,QTableView,QPushButton,QLineEdit,QHBoxLayout
from PySide6.QtGui import QStandardItemModel, QStandardItem
from PySide6.QtCore import Qt,QSortFilterProxyModel
from PySide6.QtUiTools import QUiLoader
from sqlalchemy import text
from sqlalchemy.orm import Session
from models.cc_model import CC
from models.exclude_model import Exclude
from models.database import SessionLocal

class w_cc_query(QWidget):
    def __init__(self):
        super().__init__()
        loader = QUiLoader()
        # åŠ¨æ€åŠ è½½ UI æ–‡ä»¶
        ui_path = os.path.join(os.path.dirname(__file__), '../ui/cc.ui')
        self.ui = loader.load(ui_path)

        # è®¾ç½®å¸ƒå±€
        layout = QHBoxLayout()
        layout.addWidget(self.ui)
        layout.setContentsMargins(0,0,0,0)
        self.setLayout(layout)

        # æŸ¥æ‰¾æ§ä»¶
        self.queryButton = self.ui.findChild(QPushButton,"queryButton")
        self.inputBox = self.ui.findChild(QLineEdit,"inputBox") 
        self.filterBox = self.ui.findChild(QLineEdit,"filterBox")
        self.ccView = self.ui.findChild(QTableView,"ccView")
        self.excludeView = self.ui.findChild(QTableView,"excludeView")

        self.ccView.setEditTriggers(QTableView.NoEditTriggers)
        self.excludeView.setEditTriggers(QTableView.NoEditTriggers)

        self.setup_ui()
        self.setup_table_models()  # åˆå§‹åŒ–è¡¨æ ¼æ¨¡å‹
        self.setup_connections()
    
    def setup_ui(self):
        """åˆå§‹åŒ– UI è®¾ç½®"""
        self.setWindowTitle("åŠŸèƒ½1 - å¹¶å‘ç—‡æŸ¥è¯¢")
        self.queryButton.setText("æŸ¥è¯¢")
        self.inputBox.setPlaceholderText("è¯·è¾“å…¥å¹¶å‘ç—‡ç¼–ç æŸ¥è¯¢...")
        self.filterBox.setPlaceholderText("è¾“å…¥ä¸»è¦è¯Šæ–­è¿›è¡Œè¿‡æ»¤...")
    
    def setup_connections(self):
        """è¿æ¥ä¿¡å·å’Œæ§½"""
        self.queryButton.clicked.connect(self.on_query_clicked)
        self.inputBox.returnPressed.connect(self.on_query_clicked)
        self.ccView.selectionModel().selectionChanged.connect(self.on_cc_selection_changed)
        self.filterBox.textChanged.connect(self.on_filter_text_changed)

    def setup_table_models(self):
        """åˆå§‹åŒ–è¡¨æ ¼æ•°æ®æ¨¡å‹"""
        # å·¦ä¾§è¡¨æ ¼æ¨¡å‹ - æ˜¾ç¤ºå¹¶å‘ç—‡ä¿¡æ¯
        self.cc_model = QStandardItemModel()
        self.cc_model.setHorizontalHeaderLabels(['è¯Šæ–­ç¼–ç ', 'æ’é™¤è¡¨', 'å¹¶å‘ç—‡ç±»å‹', 'CCLçº§åˆ«'])
        self.ccView.setModel(self.cc_model)
        
        # å³ä¾§è¡¨æ ¼æ¨¡å‹ - æ˜¾ç¤ºæ’é™¤æ¡ä»¶æˆ–å…¶ä»–ç›¸å…³ä¿¡æ¯
        self.exclude_model = QStandardItemModel()
        
        # æ–°å¢ï¼šåˆ›å»ºä»£ç†æ¨¡å‹ç”¨äºè¿‡æ»¤
        self.exclude_proxy_model = QSortFilterProxyModel()
        self.exclude_proxy_model.setSourceModel(self.exclude_model)
        self.exclude_proxy_model.setFilterCaseSensitivity(Qt.CaseSensitivity.CaseInsensitive)  # ä¸åŒºåˆ†å¤§å°å†™
        self.exclude_proxy_model.setFilterKeyColumn(1)  # åœ¨ç¬¬1åˆ—ï¼ˆä¸»è¦è¯Šæ–­åˆ—ï¼‰è¿›è¡Œè¿‡æ»¤
        
        self.exclude_model.setHorizontalHeaderLabels(['æ’é™¤è¡¨', 'ä¸»è¦è¯Šæ–­'])
        self.excludeView.setModel(self.exclude_proxy_model) 

        # è®¾ç½®è¡¨æ ¼å±æ€§
        self.setup_table_properties()
    
    def setup_table_properties(self):
        """è®¾ç½®è¡¨æ ¼æ˜¾ç¤ºå±æ€§"""
        # è®¾ç½®è¡¨æ ¼åˆ—å®½è‡ªé€‚åº”
        self.ccView.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        self.excludeView.horizontalHeader().setSectionResizeMode(QHeaderView.ResizeMode.Stretch)
        
        # è®¾ç½®äº¤æ›¿è¡Œé¢œè‰²
        self.ccView.setAlternatingRowColors(True)
        self.excludeView.setAlternatingRowColors(True)
        
        # è®¾ç½®é€‰æ‹©æ¨¡å¼
        self.ccView.setSelectionBehavior(QTableView.SelectionBehavior.SelectRows)
        self.excludeView.setSelectionBehavior(QTableView.SelectionBehavior.SelectRows)
    
    def on_filter_text_changed(self, text):
        """å¤„ç†filterBoxæ–‡æœ¬å˜åŒ–ï¼Œå®ç°è¿‡æ»¤åŠŸèƒ½"""
        print(f"ğŸ” è¿‡æ»¤æ–‡æœ¬å˜åŒ–: '{text}'")
        
        # è®¾ç½®è¿‡æ»¤æ¨¡å¼ä¸ºæ­£åˆ™è¡¨è¾¾å¼ï¼Œæ”¯æŒæ›´çµæ´»çš„è¿‡æ»¤
        self.exclude_proxy_model.setFilterRegularExpression(text)
        
        # æ˜¾ç¤ºè¿‡æ»¤ç»“æœä¿¡æ¯ï¼ˆå¯é€‰ï¼‰
        filtered_count = self.exclude_proxy_model.rowCount()
        total_count = self.exclude_model.rowCount()
        print(f"ğŸ“Š è¿‡æ»¤ç»“æœ: {filtered_count}/{total_count} æ¡è®°å½•")

    def on_query_clicked(self):
        """æŸ¥è¯¢æŒ‰é’®ç‚¹å‡»äº‹ä»¶"""
        code = self.inputBox.text().strip()
        if len(code) < 3:
            QMessageBox.warning(self, "è¾“å…¥é”™è¯¯", "è¯·è¾“å…¥ä¸å°äºä¸‰ä½æ•°çš„å¹¶å‘ç—‡ç¼–ç !")
            return
        print(f"æŸ¥è¯¢å¹¶å‘ç—‡: {code}")
        self.perform_query(code)
    
    def perform_query(self, code):
        """æ‰§è¡Œæ•°æ®åº“æŸ¥è¯¢æ“ä½œ"""
        db: Session = SessionLocal()
        try:
            # æŸ¥è¯¢åŒ¹é…çš„è¯Šæ–­ç¼–ç 
            cc_records = db.query(CC).filter(CC.diagcode == code).all()
            
            if cc_records:
                # æŸ¥è¯¢åˆ°è®°å½•ï¼Œæ›´æ–°è¡¨æ ¼æ˜¾ç¤º
                self.display_query_results(cc_records)
            else:
                # æœªæ‰¾åˆ°è®°å½•ï¼Œå°è¯•æ¨¡ç³ŠæŸ¥è¯¢
                self.try_fuzzy_query(db, code)
                
        except Exception as e:
            QMessageBox.critical(self, "æŸ¥è¯¢é”™è¯¯", f"æŸ¥è¯¢è¿‡ç¨‹ä¸­å‘ç”Ÿé”™è¯¯: {str(e)}")
            print(f"Database error: {e}")
            
        finally:
            db.close()
    
    def try_fuzzy_query(self, db: Session, code: str):
        """å°è¯•æ¨¡ç³ŠæŸ¥è¯¢"""
        try:
            # ä½¿ç”¨likeè¿›è¡Œæ¨¡ç³ŠæŸ¥è¯¢
            cc_records = db.query(CC).filter(CC.diagcode.like(f'{code}%')).all()
            
            if cc_records:
                self.display_query_results(cc_records)
            else:
                self.clear_table_views()
                QMessageBox.information(self, "æŸ¥è¯¢ç»“æœ", "æœªæ‰¾åˆ°ç›¸å…³è®°å½•")
                
        except Exception as e:
            QMessageBox.critical(self, "æŸ¥è¯¢é”™è¯¯", f"æ¨¡ç³ŠæŸ¥è¯¢è¿‡ç¨‹ä¸­å‘ç”Ÿé”™è¯¯: {str(e)}")
    
    def display_query_results(self, cc_records):
        """åœ¨è¡¨æ ¼ä¸­æ˜¾ç¤ºæŸ¥è¯¢ç»“æœ"""
        # æ¸…ç©ºç°æœ‰æ•°æ®
        self.cc_model.removeRows(0, self.cc_model.rowCount())
        self.exclude_model.removeRows(0, self.exclude_model.rowCount())
        
        # å¡«å……å·¦ä¾§è¡¨æ ¼ - ä¸»è¦å¹¶å‘ç—‡ä¿¡æ¯
        for record in cc_records:
            self.add_cc_record_to_table(record)
        # å¦‚æœæœ‰æ•°æ®ï¼Œé»˜è®¤é€‰æ‹©ç¬¬ä¸€è¡Œ
        # if cc_records:
        #     self.ccView.selectRow(0) 
    
    def add_cc_record_to_table(self, record: CC):
        """æ·»åŠ å•ä¸ªCCè®°å½•åˆ°å·¦ä¾§è¡¨æ ¼"""
        row = self.cc_model.rowCount()
        
        # è¯Šæ–­ç¼–ç 
        diagcode_item = QStandardItem(record.diagcode or "")
        diagcode_item.setData(record.diagcode, Qt.ItemDataRole.UserRole)  # ä¿å­˜åŸå§‹æ•°æ®
        
        # TBå­—æ®µ
        tb_item = QStandardItem(record.tb or "")
        
        # å¹¶å‘ç—‡ç±»å‹
        cctype_item = QStandardItem(record.cctype or "")
        
        # CCLç­‰çº§
        ccl_item = QStandardItem(str(record.ccl) if record.ccl is not None else "")
        
        # æ ¹æ®CCLç­‰çº§è®¾ç½®ä¸åŒçš„èƒŒæ™¯è‰²
        if record.ccl is not None:
            if record.ccl == 2:
                ccl_item.setBackground(Qt.GlobalColor.red)
                ccl_item.setForeground(Qt.GlobalColor.white)
            elif record.ccl == 1:
                ccl_item.setBackground(Qt.GlobalColor.green)
                ccl_item.setForeground(Qt.GlobalColor.black)
        
        # æ·»åŠ åˆ°æ¨¡å‹
        self.cc_model.appendRow([diagcode_item, tb_item, cctype_item, ccl_item])
    
    def on_cc_selection_changed(self, selected, deselected):
        """å·¦ä¾§è¡¨æ ¼é€‰æ‹©å˜åŒ–äº‹ä»¶"""
        print(f"ğŸ” é€‰æ‹©å˜åŒ–äº‹ä»¶è§¦å‘")
        print(f"   é€‰ä¸­çš„ç´¢å¼•: {selected.indexes()}")
        print(f"   å–æ¶ˆé€‰ä¸­çš„ç´¢å¼•: {deselected.indexes()}")
        
        # è·å–å½“å‰é€‰ä¸­çš„è¡Œ
        selected_indexes = self.ccView.selectionModel().selectedRows()
        print(f"   é€‰ä¸­çš„è¡Œ: {[index.row() for index in selected_indexes]}")
        
        if not selected_indexes:
            print("   âŒ æ²¡æœ‰é€‰ä¸­ä»»ä½•è¡Œ")
            self.clear_exclude_table()
            return
        
        # è·å–ç¬¬ä¸€é€‰ä¸­çš„è¡Œ
        selected_row = selected_indexes[0].row()
        print(f"   å¤„ç†ç¬¬ {selected_row} è¡Œ")
        
        # æ£€æŸ¥æ¨¡å‹æ•°æ®
        print(f"   æ¨¡å‹è¡Œæ•°: {self.cc_model.rowCount()}")
        print(f"   æ¨¡å‹åˆ—æ•°: {self.cc_model.columnCount()}")
        
        # å®‰å…¨åœ°è·å–åˆ—æ•°æ®
        try:
            # è·å–è¯¥è¡Œçš„æ‰€æœ‰åˆ—æ•°æ®
            for col in range(self.cc_model.columnCount()):
                index = self.cc_model.index(selected_row, col)
                if index.isValid():
                    data = self.cc_model.data(index)
                    user_data = self.cc_model.data(index, Qt.ItemDataRole.UserRole)
                    print(f"   åˆ— {col}: æ˜¾ç¤ºæ•°æ®='{data}', ç”¨æˆ·æ•°æ®='{user_data}'")
                else:
                    print(f"   åˆ— {col}: ç´¢å¼•æ— æ•ˆ")
            
            # è·å–è¯¥è¡Œçš„tbå€¼ï¼ˆç¬¬1åˆ—ï¼‰
            tb_index = self.cc_model.index(selected_row, 1)  # ç¬¬1åˆ—æ˜¯tbå­—æ®µ
            if tb_index.isValid():
                tb_value = self.cc_model.data(tb_index, Qt.ItemDataRole.UserRole)
                print(f"   TBå€¼ (UserRole): '{tb_value}'")
                
                # å¦‚æœUserRoleæ²¡æœ‰å€¼ï¼Œå°è¯•è·å–æ˜¾ç¤ºæ–‡æœ¬
                if not tb_value:
                    tb_value = self.cc_model.data(tb_index)  # è·å–æ˜¾ç¤ºæ–‡æœ¬
                    print(f"   TBå€¼ (æ˜¾ç¤ºæ–‡æœ¬): '{tb_value}'")
                
                if tb_value:
                    print(f"   ğŸ“Œ å¼€å§‹æŸ¥è¯¢æ’é™¤è¡¨ï¼ŒTBå€¼: {tb_value}")
                    self.query_exclude_by_tb(tb_value)
                else:
                    print("   âŒ æ²¡æœ‰è·å–åˆ°TBå€¼")
                    self.clear_exclude_table()
            else:
                print("   âŒ TBç´¢å¼•æ— æ•ˆ")
                self.clear_exclude_table()
                
        except Exception as e:
            print(f"   âŒ è·å–æ•°æ®æ—¶å‡ºé”™: {e}")
            import traceback
            print(traceback.format_exc())
            self.clear_exclude_table()
    
    def query_exclude_by_tb(self, tb_value):
        """æ ¹æ®tbå€¼æŸ¥è¯¢Excludeè¡¨"""
        db: Session = SessionLocal()
        try:
            # æŸ¥è¯¢Excludeè¡¨ä¸­åŒ¹é…çš„è®°å½•
            exclude_records = db.query(Exclude).filter(Exclude.tb == tb_value).all()
            
            # æ›´æ–°å³ä¾§è¡¨æ ¼
            self.update_exclude_table(exclude_records)
            
            print(f"ğŸ“Š æ‰¾åˆ° {len(exclude_records)} æ¡æ’é™¤è®°å½•")
            
        except Exception as e:
            QMessageBox.critical(self, "æŸ¥è¯¢é”™è¯¯", f"æŸ¥è¯¢æ’é™¤è¡¨æ—¶å‘ç”Ÿé”™è¯¯: {str(e)}")
            print(f"Exclude query error: {e}")
            
        finally:
            db.close()  
    
    def update_exclude_table(self, exclude_records):
        """æ›´æ–°å³ä¾§æ’é™¤è¡¨æ ¼"""
        # æ¸…ç©ºç°æœ‰æ•°æ®
        self.exclude_model.removeRows(0, self.exclude_model.rowCount())
        
        self.filterBox.clear()

        if exclude_records:
            # å¡«å……æ’é™¤æ•°æ®
            for record in exclude_records:
                tb_item = QStandardItem(record.tb or "")
                maindiag_item = QStandardItem(record.maindiag or "")
                
                self.exclude_model.appendRow([tb_item, maindiag_item])
        else:
            # æ²¡æœ‰æ’é™¤è®°å½•
            no_data_item = QStandardItem("æ— æ’é™¤æ•°æ®")
            no_data_item.setTextAlignment(Qt.AlignmentFlag.AlignCenter)
            self.exclude_model.appendRow([no_data_item, QStandardItem("")])    
    
    def clear_exclude_table(self):
        """æ¸…ç©ºå³ä¾§è¡¨æ ¼"""
        self.exclude_model.removeRows(0, self.exclude_model.rowCount())
        no_data_item = QStandardItem("è¯·é€‰æ‹©å·¦ä¾§è¡Œ")
        no_data_item.setTextAlignment(Qt.AlignmentFlag.AlignCenter)
        self.exclude_model.appendRow([no_data_item, QStandardItem("")])

        self.filterBox.clear()
    
    def clear_table_views(self):
        """æ¸…ç©ºè¡¨æ ¼æ˜¾ç¤º"""
        self.cc_model.removeRows(0, self.cc_model.rowCount())
        self.exclude_model.removeRows(0, self.exclude_model.rowCount())
        
        self.filterBox.clear()

        # æ·»åŠ æç¤ºä¿¡æ¯
        no_data_item = QStandardItem("æ— æ•°æ®")
        no_data_item.setTextAlignment(Qt.AlignmentFlag.AlignCenter)
        self.cc_model.appendRow([no_data_item, QStandardItem(""), QStandardItem(""), QStandardItem("")])
        
        self.exclude_model.appendRow([QStandardItem("æ— "), QStandardItem("æœªæ‰¾åˆ°åŒ¹é…è®°å½•")])
    
    def closeEvent(self, event):
        """çª—å£å…³é—­äº‹ä»¶ï¼Œç”¨äºæ¸…ç†èµ„æº"""
        super().closeEvent(event)